deploy/crds/operators.coreos.com_servicebindings_crd.yaml

```diff
@@ -275,9 +230,37 @@ spec:
             mountPathPrefix:
               description: MountPathPrefix is the prefix for volume mount
               type: string
+            services:
```

Comment:
Looks like this is a major gap between the spec and our implementation. The spec only supports only one `service` per `ServiceBinding` CRD. https://k8s-service-bindings.github.io/spec/#resource-type-schema-1
@sbose78 What would you suggest here?

Reply:
We would lose critical use cases with a single `service` and hence we'll stick to `services`. The CRD/API evolution will happen within the project and not **necessarily** as per the spec. 

After we reach a reasonable state of stability ( next upstream ) release, we'll use the enhancement proposal process to make well socialized and thought our changes from our existing state.

Reply:
I would suggest not to use the word `ServiceBinding` to avoid confusion with the spec. This is not going to help our users. Let's stick with `ServiceBindingRequest` and continue with rest of the changes.

If the spec is going to co-exist in the Kubernetes community, it will be really confusing to our end-users. Sometimes they might read the spec and expect the same API here as well.

Reply:
I can think of a couple more practical reason to not use `ServiceBinding`:

1. Potential future compatibility: If we choose later to adapt the spec, it would be easy as it will be new CRD
2. Community members can create some kind of bridge between the objects of spec CRD and our CRD.

Reply:
Would our API group being different make things any better, in your opinion ?

Note, we are never going to move into a new set of CRDs at one go, we'll continue to make incremental changes.
https://github.com/redhat-developer/service-binding-operator/issues/587#issuecomment-671682260

> If the spec is going to co-exist in the Kubernetes community, it will be really confusing to our end-users. Sometimes they might read the spec and expect the same API here as well.

I did consider this, however, since these APIs are not part of core kubernetes, there's no way a user would read the spec and want to try things out on their cluster. 

Users have to make a conscious choice to install a specific controller+CRD / operator. That should 'tell' them what to do next and which CRs to create next. The "user documentation" should really come from implementation and not the spec.

Reply:
The spec suggests using multiple ServiceBinding resources to bind multiple services to a single application. What value does the the single SBRs support for multiple services offer over simply using multiple SBs each of which references the same application?

Reply:
Ron, good question!

Here's why we had gone down the path of multiple services:

1. In a lot of cases, to achieve anything reasonable, multiple services need to be used ( CR, secret, configmap ), unless the backing service authors had made code changes to include everything interesting in the same CR. This project aims to allow backing service authors to get away with absolutely bare minimum changes.

2. Having multiple objects ( SBs ) when you can have one feels like an overkill to me especially when you have a controller reconciling on those and modifying a workload as part of that.

Reply:
1. Everyone agrees that multiple services exist in the real world.    But that's not affecting the creators of backing services.
2. When I asked about the second point, this is the response I got from @scothis
   "multiple services would be much more complex to support. The `type`, `provisioner`, `mappings` and `env` are all specific to a single service" (Ref: https://kubernetes.slack.com/archives/C012F2GPMTQ/p1598966672013000?thread_ts=1598960730.010900&cid=C012F2GPMTQ )

Reply:
There are a couple different approaches to support this behavior that are perfectly acceptable to the spec:
1. decompose a higher order resource into multiple `ServiceBinding` resources
2. create an alternative equivalent to the `ServiceBinding` resource that consumes the [Provisioned Service](https://github.com/k8s-service-bindings/spec#provisioned-service) and [Application Projection](https://github.com/k8s-service-bindings/spec#application-projection) pillars of the spec independently.

The primary intent of the spec is to provide a baseline set of behavior that is portable across implementations. Each implementation is free to go beyond the spec and provide high level or more capable behavior, at the cost of no longer being portable. You can choose the right experience to offer to your users and customers.

Reply:
> Everyone agrees that multiple services exist in the real world.

@baijum , 
Note, I was talking about one instance of a database service which gets represented as multiple "service binding" spec.services here. 

Example, a single PostgreSQL database service might be represented as 
* one CR, 
* one secret, 
* one ConfigMap  .. which aren't necessarily referenced in the CR

I wouldn't want to have the user create multiple SBs with each of the above objects listed as a `spec.service`.

> But that's not affecting the creators of backing services.

It does impact the backing service author if we ask the backing service author to ensure everything associated with one instance of a database is present in one CR.

I'm not making an argument against the specification's proposal, rather I'm arguing why we considered multiple backing services in the first place in a single `ServiceBinding[Request]`

Reply:
> Note, I was talking about one instance of a database service which gets represented as multiple "service binding" spec.services here.

This concept was in RC1, but was removed for RC2 after a good bit of discussion. Forcing a user to be aware of multiple discrete parts that compose to represent a single service is not a great experience. In an ideal world, it's better to push that problem to the service provider to provide all the metadata a user needs to consume a service within a single package.

Knowing that we don't live in an ideal world, for the case where supporting existing resources is necessary, the ServiceBinding `mappings` are available to inject additional values or the [Binding Secrete Generation Strategies](https://github.com/k8s-service-bindings/spec#binding-secret-generation-strategies) spec extension. For even more advanced needs a new custom resource could aggregate complex state and expose a uniform [Provisioned Service(https://github.com/k8s-service-bindings/spec#provisioned-service).

Reply:
> Forcing a user to be aware of multiple discrete parts that compose to represent a single service is not a great experience

Not saying you said so :) But, "Service Binding" isn't the one really forcing that. The provisioned service forces that.

If a single package is available, spec.services[0] is all we need. As you mentioned, in an ideal world, provisioned services should be the ones making things easily consumable as a single package. We don't control the world of provisioned services, but we can make it a little easy to use existing provisioned services in service binding scenarios.

>  the ServiceBinding mappings are available to inject additional values or the Binding Secrete Generation Strategies spec extension

Unfortunately, these wouldn't work unless the discrete parts are referenced in the backing service CR's spec or status.

> For even more advanced needs a new custom resource could aggregate complex state and expose a uniform [Provisioned Service(https://github.com/k8s-service-bindings/spec#provisioned-service).

We all agree that there is value in aggregation. There are different ways to get there. Remember this project didn't have multiple services to begin with :) 

@baijum @scothis 

I'm not necessarily giving an opinion on the `ProvisionedService` , if you feel it is valuable, please create an enhancement proposal ( which can be as simple as a Github issue ) to begin with. We do need to have a path to go from existing implementation to the suggested one and therefore a proposal with details would be awesome.

Reply:
just to add my 2 cents :)  I used to also really like specifying multiple services on the same `ServiceBinding`, but two points that made me agree with a single service per `ServiceBinding` were:
* it's easier for a tool to add and remove services by simply adding or removing independent CRs, instead of having to deal with editing (and potentially corrupting, if concurrent updates are not well synchronized) an array inside a CR. It simply becomes "kubectl apply .." and "kubectl delete .."
* cleaner association of attributes with the service.  For example,  when we identify a set of extra mappings or environment variables in a `ServiceBinding`, which service does that get associated with if we have multiple services?  Or if we want to change the service name or type that gets mounted, we again run into the issue that they all require a way to identify the service you want to apply those attributes against.